File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/codesigning/elf/ElfFile.java
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.codesigning.elf;

import com.ohos.hapsigntool.codesigning.exception.ElfFormatException;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * ELF file
 *
 * @since 2024/07/01
 */
public class ElfFile {
    private ElfHeader elfHeader;

    private final List<ElfProgramHeader> programHeaderList = new ArrayList<>();

    /**
     * Constructor for ElfFile
     *
     * @param is InputStream
     * @throws IOException io error
     * @throws ElfFormatException elf file format error
     */
    public ElfFile(InputStream is) throws IOException, ElfFormatException {
        elfHeader = new ElfHeader(is);
        if (!isElfFile()) {
            return;
        }
        byte eiClass = elfHeader.getEiClass();
        byte eiData = elfHeader.getEiData();
        int ePhnum = elfHeader.getEPhnum();
        long ePhOff = elfHeader.getEPhOff();
        if (eiClass == ElfDefine.ELF_32_CLASS) {
            is.skip(ePhOff - ElfDefine.ELF_HEADER_32_LEN);
        } else if (eiClass == ElfDefine.ELF_64_CLASS) {
            is.skip(ePhOff - ElfDefine.ELF_HEADER_64_LEN);
        }
        for (int i = 0; i < ePhnum; i++) {
            ElfProgramHeader pHeader = new ElfProgramHeader(is, eiClass, eiData);
            programHeaderList.add(pHeader);
        }
    }

    /**
     * filter executable program segment headers
     *
     * @return executable program segment headers
     */
    public List<ElfProgramHeader> filterExecPHeaders() {
        return programHeaderList.stream().filter(phdr -> (phdr.getPFlags() & 1) == 1).collect(Collectors.toList());
    }

    /**
     * return true if magic number is correct
     *
     * @return true if magic number is correct
     */
    public final boolean isElfFile() {
        return elfHeader.isElfFile();
    }
}

File: OpenHarmony/developtools/integration_verification/tools/deps_guard/elf_file_mgr/elf_file.py
#!/usr/bin/env python
#coding=utf-8

#
# Copyright (c) 2022 Huawei Device Co., Ltd.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import os
from stat import ST_SIZE

from .utils import command


class ElfFile(dict):
    def __init__(self, file, prefix):
        self._f = file
        self._f_safe = "'%s'" % file

        self["name"] = os.path.basename(file)
        self["size"] = os.stat(self._f)[ST_SIZE]
        if self["name"].find(".so") > 0:
            self["type"] = "lib"
        else:
            self["type"] = "bin"
        self["path"] = file[len(prefix):]

    def __eq__(self, other):
        if not isinstance(other, ElfFile):
            return NotImplemented

        return self["path"] == other["path"]

    def is_library(self):
        if self["name"].find(".so") > 0:
            return True
        return False

    def get_file(self):
        return self._f

    # Return a set of libraries the passed objects depend on.
    def library_depends(self):
        if not os.access(self._f, os.F_OK):
            raise Exception("Cannot find lib: " + self._f)
        dynamics = command("readelf", "--dynamic", self._f_safe)
        res = []
        for line in dynamics:
            pos = line.find("(NEEDED)")
            if pos <= 0:
                continue
            line = line[pos + 8:]
            line = line.strip()
            if not line.startswith("Shared library:"):
                continue
            line = line[15:]
            line = line.strip()
            if line.startswith("["):
                line = line[1:]
            if line.endswith("]"):
                line = line[:-1]
            line = line.strip()
            res.append(line)
        return res

    def __extract_soname(self):
        soname_data = command("mklibs-readelf", "--print-soname", self._f_safe)
        if soname_data:
            return soname_data.pop()
        return ""

    def __extract_elf_size(self):
        size_data = command("size", self._f_safe)
        if not size_data or len(size_data) < 2:
            self["text_size"] = 0
            self["data_size"] = 0
            self["bss_size"] = 0
            return 0

        vals = size_data[1].split()
        self["text_size"] = int(vals[0])
        self["data_size"] = int(vals[1])
        self["bss_size"] = int(vals[2])
        return ""

if __name__ == '__main__':
    import elf_walker

    cnt = 0
    elfFiles = elf_walker.ELFWalker()
    for f in elfFiles.get_elf_files():
        if f.find("libskia_ohos.z.so") < 0:
            continue
        elf = ElfFile(f, elfFiles.get_product_images_path())
        print(f)
